internal/femto/femto.go

```diff
@@ -0,0 +1,68 @@
+// Package femto provides a tiny abstraction layer for writing API's that serve JSON over HTTP.
+package femto
+
+import (
+	"encoding/json"
+	"log/slog"
+	"net/http"
+	"sync/atomic"
+)
+
+type Femto struct {
+	mux   http.ServeMux
+	reqNo atomic.Uint64
+}
+
+func (f *Femto) ListenAndServe(addr string) error {
+	return http.ListenAndServe(addr, &f.mux)
+}
+
+func OnPost[T any](f *Femto, pattern string, handle HandlerFunc[T]) {
+	f.mux.HandleFunc(pattern, func(w http.ResponseWriter, r *http.Request) {
+		doPost(f, w, r, handle)
+	})
+}
+
+func doPost[T any](f *Femto, w http.ResponseWriter, r *http.Request, handle HandlerFunc[T]) {
+
+	reqNo := f.nextReqNo()
+	log := f.logger().With(slog.Uint64("req_no", reqNo))
+
+	log.Info("New Request", "method", r.Method, "url", r.URL, "from", r.RemoteAddr)
+
+	if r.Method != http.MethodPost {
+		log.Info("Wanted GET, returned 405")
+		http.Error(w, "Expected GET", http.StatusMethodNotAllowed)
+		return
+	}
+
+	// TODO: Ensure all field's are present. I'm so mad this is hard.
```

Comment:
I think checking that all fields are present should be okay to do - check my PR on #28

Reply:
that linked PR doesn't do this:

```go
func TestBuildStatusObjectEmpty(t *testing.T) {
	t.Parallel()

	emptyObj := []byte("{}")
	obj, err := buildStatusObject(emptyObj)
	assert.NoError(t, err)
	assert.Zero(t, obj)
}
```

passes (and it shouldn't)

Reply:
Huh, annoying. I saw this edge case on toml deserialisation too, although I fixed that. Seems like this is an edge case that can be manually guarded against in buiildStatusObject if this is the only thing that needs to be checked for.

With that in mind, why on earth does deserialisation fail on all other field mismatch cases but not the empty case? These semantics are weird...

Reply:
it also passes on the missing fields case:

```go
func TestBuildStatusObjectMissingFields(t *testing.T) {
	t.Parallel()

	emptyObj := []byte(`{"gpu_name": "foo"}`)
	obj, err := buildStatusObject(emptyObj)
	assert.NoError(t, err)
	assert.Equal(t, "foo", obj.Name)
}
```

passes

Reply:
`TestBuildMalformedObject` passes because we have ``DriverVersion string `json:"driver_ver"``, but in JSON it's a number: `"driver_ver": 1.0`

so the error is `json: cannot unmarshal number into Go struct field GPUStatusPacket.driver_ver of type string`

the unmarhall failed, but not due to missing fields. i guess we should assert on the errors contents, not just that it's non-nil

Reply:
That makes more sense, that's definitely worth looking into then. I'll make sure to fix that and have some interface for deserialisation that checks for these failures.